dap-launch.el

```diff
@@ -0,0 +1,241 @@
+                              dir))
+
+(defun dap--buffer-basename ()
+  "Return the name of the current buffer's file without its directory."
+  (file-name-nondirectory buffer-file-name))
+
+(defun dap--buffer-basename-sans-extension ()
+  "Same as `dap--buffer-basename', but without the extension."
+  (file-name-sans-extension (dap--buffer-basename)))
+
+(defun dap--buffer-extension ()
+  "Return the extension of the buffer's file with a leading dot.
+If there is either no file associated with the current buffer or
+if that file has no extension, return the empty string."
+  (if-let ((buffer-name buffer-file-name)
+           (ext (file-name-extension buffer-name)))
+      (concat "." ext)
+    ""))
+
+(defun dap--buffer-dirname ()
+  "Return the directory the buffer's file is in."
+  (file-name-directory buffer-file-name))
+
+(defun dap--buffer-current-line ()
+  "Return the line the cursor is on in the current buffer."
+  (number-to-string (line-number-at-pos)))
+
+(defun dap--buffer-selected-text ()
+  "Return the text selected in the current buffer.
+If no text is selected, return the empty string."
+  ;; Cannot fail, as if there is no mark, (mark) and (point) will be equal, and
+  ;; (`buffer-substring-no-properties') will yield "", as it should.
+  (buffer-substring-no-properties (mark) (point)))
+
+(defun dap--launch-json-warn-nil (text)
+  "Issue a warning related to the launch.json file containing TEXT.
+Always return nil."
+  (message (concat "warning: launch.json: " text))
+  nil)
+
+(defun dap--warn-unknown-envvar (var)
+  "Warn, related to launch.json, that VAR is an unknown environment variable.
+Always return nil."
+  (dap--launch-json-warn-nil (format "no such environment variable '%s'" var)))
+
+(defun dap--launch-json-getenv (var)
+  "Return the environment variable in matched string VAR.
+Only for use in `dap-launch-json-variables'."
+  (let ((envvar (match-string 1 var)))
+    (or (getenv envvar) (dap--warn-unknown-envvar envvar) "")))
+
+(defvar dap-launch-json-variables
+  ;; list taken from https://code.visualstudio.com/docs/editor/variables-reference
+  '(("workspaceFolderBasename" . dap--project-project-basename)
+    ("workspaceFolder" . lsp-workspace-root)
+    ("relativeFileDirname" . dap--project-relative-dirame)
+    ("relativeFile" . dap--project-relative-file)
+    ("fileBasenameNoExtension" . dap--buffer-basename-sans-extension)
+    ("fileBasename" . dap--buffer-basename)
+    ("fileDirname" . dap--buffer-dirname)
+    ("fileExtname" . dap--buffer-extension)
+    ("lineNumber" . dap--buffer-current-line)
+    ("selectedText" . dap--buffer-selected-text)
+    ("file" . buffer-file-name)
+    ("env:\\(.*\\)" . dap--launch-json-getenv)
+    ;; technically not in VSCode, but I still wanted to add a way to escape $
+    ("$" . "$")
+    ;; the following variables are valid in VSCode, but have no meaning in
+    ;; Emacs, and are as such unsupported.
+    ;; ("cwd") ;; the task runner's current working directory,
```

Comment:
This is a bit of a misnomer by microsoft. "cwd" has nothing to do with the program's current working directory, but is in fact "the task runner's current working directory" (see here: https://code.visualstudio.com/docs/editor/variables-reference, also here https://code.visualstudio.com/docs/editor/tasks for what a task runner is). Since there is no dap-mode task runner, this variable does not make sense.

Reply:
Thank you for clearing that up. As a side note, I am planning to create tasks extensions for emacs.

Reply:
I will add it as a variable if its done then.